blockstore/apps/bundles/models.py

```diff
@@ -0,0 +1,223 @@
+of what's inside the Bundles is handled through the bundles.store module.
+
+# Scaling Considerations
+
+Taking edx.org as measuring stick, we want to be able to easily support:
+* 10K+ courses
+* 10K+ content libraries
+* ~20 Sequences in each course on average, but allowing for a max of 100
+* ~50+ TB of data (mostly video)
+* 10s to 100s of millions of Content Bundles
+* ~20-30 versions per Bundle on average
+* Up to 100 files per bundle
+
+# MySQL Notes
+
+Note that while this data model should work in MySQL or PostgreSQL, much of it
+is specifically designed around making MySQL work well at scale. Open edX by
+default uses MySQL for historical reasons (PostgreSQL wasn't offered on AWS RDS
+at the time edx-platform was created), and we're not looking to introduce the
+operational overhead of requiring another data store.
+
+But this will make the schema look a bit weird. Some MySQL-isms:
+
+1) MySQL insert performance breaks down if you use UUIDs for a primary key. The
+primary key is used as the clustering index, meaning that MySQL will store table
+data in primary key order. If that ordering is random, your data is much more
+fragmented. For that reason, when UUIDs are needed, they're used as secondary
+unique indexes. Those indexes may get fragmented, but it will have less of an
+impact overall, and we're never going to do range queryies on UUIDs anyhow. Side
+note: If you use UUID1 and get fancy with bit re-ordering, you can get around
+this, but it adds complexity to the code and prevents us from using the randomly
+generated UUID4.
+
+2) Another reason we're keeping UUIDs out of the primary key is because all
+secondary keys in MySQL's InnoDB storage engine store a copy of the primary key
+entry as well. A bigint is 8 bytes, and even an optimally packed UUID is still
+16 bytes.
+
+3) Our repos generally have VARCHAR(255) everywhere because MySQL's "utf8"
+encoding uses a max of three bytes per character (note: meaning it's not really
+UTF-8), and 255 * 3 = 765 bytes. The largest size for any column in an InnoDB
+index is 767 bytes. However, if we use MySQL's "utf8mb4", a.k.a. "actually
+UTF-8", that limit goes down to 191 bytes (191 * 4 = 764). Because 191 is a
+peculiar and difficult to remember number for an API, I am somewhat arbitrarily
+using 180 as a reasonable sounding number below the limit.
+
+4) There are a number of models that are made bigint when they don't strictly
+need to be (Bundles are unlikely to grow to that size, for instance), because
+it's not *that* expensive, and migrating data once it gets that large is a pain.
+"""
+import codecs
+import uuid
+
+from django.db import models
+from django.dispatch import receiver
+import attr
+
+from .store import BundleDataStore, snapshot_created
+
+MAX_CHAR_FIELD_LENGTH = 180
+
+
+class Collection(models.Model):
+    """
+    Administrative grouping for Bundles: policy, permissions, licensing, etc.
```

Comment:
I've shared a folder with you all containing some discussions done among a representative of Creative Commons, edX's OSCM at the time, a community member, and OpenCraft, which might help clarify some of the use cases requested for CC licensing.

(One of the topics that came up during those discussions was the desire for exporting/importing partial courses, which, of course, Blockstore achieves beautifully 😄 )

Having the ability to say "all versions of the Bundles in this Collection are now available under Creative Commons" gets tricky, if the license becomes more restrictive (e.g. adds a non-commercial clause), and existing links to the more open content bundles already exist.  So I think this means we have to keep track of the license of a particular BundleVersion, and notify any linked content of license changes prior to receiving updates.

>  It would also allow us to handle cases where things are licensed in multiple ways

Which "things"?  Since we're saying a Bundle is a discrete set of learning content, then I think it's fair to say that the Bundle will have a license that applies to everything it contains (for a given version or set of versions).  And so the license of the Collection itself is an amalgam of the licenses of its Bundles.  Whether or not the Collection has its own license -- don't know if is necessary or not.

Reply:
> Which "things"? Since we're saying a Bundle is a discrete set of learning content, then I think it's fair to say that the Bundle will have a license that applies to everything it contains (for a given version or set of versions). And so the license of the Collection itself is an amalgam of the licenses of its Bundles. Whether or not the Collection has its own license -- don't know if is necessary or not.

Sorry, that was really unclear of me. I was thinking that if a Collection (and by extension a Bundle) were available under multiple licenses simultaneously (e.g. GPL/BSD), we'd want some way to identify which license was being used on any particular borrowing. I agree that all the contents of a Bundle should share the same license. I would also like to take the tack that the entire Collection shares the same license.

> Having the ability to say "all versions of the Bundles in this Collection are now available under Creative Commons" gets tricky, if the license becomes more restrictive (e.g. adds a non-commercial clause), and existing links to the more open content bundles already exist. So I think this means we have to keep track of the license of a particular BundleVersion, and notify any linked content of license changes prior to receiving updates.

I'm leaning towards capturing it at the Link level, so we'd only conceptually have two things: what licenses the Collection is available under now (captured at the Collection level), and what license was this particular BundleVersion when you decided to borrow it into your own Bundle (attached to the Link model). In the case where licensing gets more restrictive, we'd still be able to look up all Bundles that borrowed under the more permissive license and do something about it if necessary. Otherwise, they would be allowed to use the old thing but wouldn't be allowed to upgrade.

This gives Links and Licensing a tighter relationship than in the original design, but I think that's okay, and makes sense if we're going to treat licensing seriously.

Reply:
> I'm leaning towards capturing it at the Link level, so we'd only conceptually have two things: what licenses the Collection is available under now (captured at the Collection level), and what license was this particular BundleVersion when you decided to borrow it into your own Bundle (attached to the Link model).

That sounds good to me.

> In the case where licensing gets more restrictive, we'd still be able to look up all Bundles that borrowed under the more permissive license and do something about it if necessary. Otherwise, they would be allowed to use the old thing but wouldn't be allowed to upgrade.

That would require that Blockstore "understand" the various licenses and their restrictions, right?

Reply:
> I've shared a folder with you all containing some discussions done among a representative of Creative Commons, edX's OSCM at the time, a community member, and OpenCraft, which might help clarify some of the use cases requested for CC licensing.

Thanks! That is great. Will give it a read.

>  Since we're saying a Bundle is a discrete set of learning content, then I think it's fair to say that the Bundle will have a license that applies to everything it contains (for a given version or set of versions).

A bundle may have text, javascript libs, images and videos taken from other places on the internet with different licenses. I imagine that would be very common. What is the license of the bundle in that case?

> I'm leaning towards capturing it at the Link level, so we'd only conceptually have two things: what licenses the Collection is available under now (captured at the Collection level), and what license was this particular BundleVersion when you decided to borrow it into your own Bundle (attached to the Link model).

This breaks the common convention that the license of a version of an item is what was declared at the time it was released. If I am linking to bundle1@1 from bundle2 I should be able link to it from bundle3 even if bundle1@2 has a more restrictive license.

> I think this means we have to keep track of the license of a particular BundleVersion, and notify any linked content of license changes prior to receiving updates.

This makes more sense to me.

I wasn't part of the early discussions but my feeling is that the goal of tracking license info as a first class thing is to improve discovery of items for re-use (and to enforce restrictions) e.g. "this bundle can be used without modification", "this bundle requires attribution for re-use". I think it would be better if we first came up with a list of scenarios/use-cases that would need to be handled and then figure out what data model would work best. Makes sense?

Reply:
> That would require that Blockstore "understand" the various licenses and their restrictions, right?

Yeah. I'm still hoping the layer that knows is relatively thin and isolated, but yeah.

> A bundle may have text, javascript libs, images and videos taken from other places on the internet with different licenses. I imagine that would be very common. What is the license of the bundle in that case?

I don't think it's worth the cost to try to track every JavaScript library file in a Bundle and map out their licenses. Blockstore can support a handful of pre-baked licenses at the Collection level -- probably "all rights reserved, only my university gets to use this", and the various flavors of Creative Commons. If a user is going to put a separately licensed third-party piece of content into their Bundle, it should be their responsibility to make sure the licensing is compatible. It should also be their responsibility to include any information needed for proper attribution of authorship.

> This breaks the common convention that the license of a version of an item is what was declared at the time it was released. If I am linking to bundle1@1 from bundle2 I should be able link to it from bundle3 even if bundle1@2 has a more restrictive license.

As a developer mapping this onto the source code analogy, I agree. But I don't think that necessarily maps with product expectations. I looked through the docs @pomegranited shared, and I can't see any place this was hashed out.

My thinking was basically:

1. Going from a more permissive to more restrictive license is likely going to be an edge case.
2. If it's really a massive shift-over, it's quite likely enough to warrant the creation of a new Collection.
3. It's just as likely (or more likely) going to be an accident. As in "we originally put this as CC, but we meant to be open with that _other_ Collection, and not this one". We'll need to handle this case anyway.
4. We do need to keep track of license at time of borrowing so as to not completely screw over the people who used it on that basis...
5. But to keep track of it on a per-BundleVersion level seems like it would introduce a lot of complexity to support a rare operation.

As an example of complexity, let's take discovery. If we do licensing all at the Collection level, then for search purposes something is either public because the Collection is permissively licensed, or it's closed off from browsing unless you belong to that institution. If we do it at the BundleVersion layer, do we make just those specific versions discoverable in search? It really might as well not exist for borrowing with that permissive license if nobody can find it anyway... but then do we work that into the UI so that we're now even seeing subsets of the appropriately licensed versions for an individual Bundle? What does the UI look like for that? What does it look like if the license has changed a couple of times?

It just seems like potentially a lot of edge cases for relatively little gain. Though maybe I'm just not modeling it right in my head...?

> I wasn't part of the early discussions but my feeling is that the goal of tracking license info as a first class thing is to improve discovery of items for re-use (and to enforce restrictions) e.g. "this bundle can be used without modification", "this bundle requires attribution for re-use". I think it would be better if we first came up with a list of scenarios/use-cases that would need to be handled and then figure out what data model would work best and what support needs to be added to the permissions system. Makes sense?

Agreed.

Reply:
> I don't think it's worth the cost to try to track every JavaScript library file in a Bundle and map out their licenses. Blockstore can support a handful of pre-baked licenses at the Collection level -- probably "all rights reserved, only my university gets to use this", and the various flavors of Creative Commons. If a user is going to put a separately licensed third-party piece of content into their Bundle, it should be their responsibility to make sure the licensing is compatible. It should also be their responsibility to include any information needed for proper attribution of authorship.

Makes sense. So basically it must be possible for all content added to a bundle or collection to be offered under the license that is attached to the bundle or collection (depending on which level the license is being attached). 

> As an example of complexity, let's take discovery. If we do licensing all at the Collection level, then for search purposes something is either public because the Collection is permissively licensed, or it's closed off from browsing unless you belong to that institution. If we do it at the BundleVersion layer, do we make just those specific versions discoverable in search? It really might as well not exist for borrowing with that permissive license if nobody can find it anyway... but then do we work that into the UI so that we're now even seeing subsets of the appropriately licensed versions for an individual Bundle? What does the UI look like for that? What does it look like if the license has changed a couple of times?

I am wondering if it would make more sense for authors to specify the policies first and then they are given the option to specify a license label to display that matches those policies. The bundle versions can have the license label attached to them (at the time of creation). However, what we associate with the links are going to be policy objects. This way there can be restrictions on some kinds of policy transitions and it would be simpler to deal with the others.

I am not sure it would work legally to move version 10 of a bundle to a different license (which is not a superset of the earlier license) and then have that license label be displayed when an earlier version is viewed. However, it should be perfectly okay for authors to change (including restrict) the policies around discovery and reuse of all the versions of a bundle or collection in the Blockstore (you are of course free to export or copy an earlier version if you have access to it via a link created before and the license allows it; there is no reliable DRM anyway). I do agree that overall it would be much simpler if these policies applied to all the versions.

I have also created a discovery task for this: https://tasks.opencraft.com/browse/LX-51